### PR TITLE
Oas

### PR DESCRIPTION
**Pull request by Dashron**
This is ready for review. You can see the oas file in the root directory under `openapi-schema.yml`, and you can see an example of it rendered on readme.com here: https://dashron-o5etest.readme.io/reference/listmanifests.

- I've added a system to augment data on each endpoint
- I've added a system to post-process the oas file to add or remove data as you see fit
- I've added documentation on how to generate a new OAS file

### CODE REVIEW

File: api/serializers.py [hunk 1]
```diff
@@ -34,8 +36,8 @@ class DocumentSerializer(serializers.HyperlinkedModelSerializer):
     class Meta:
             model = Document
             fields = (
-                'title', 
-                'slug', 
+                'title',
```
Comment (Dashron) [inline api/serializers.py:39]: these are all removed automatically from my IDE. I managed to ignore them later, but missed them in these early commits. Let me know if you want me to bring them back.

File: server/settings.py [hunk 2]
```diff
@@ -30,7 +30,7 @@
 DEBUG = os.environ.get('OPEN5E_DEBUG', '') != 'False'
 
 ## Always allow connections from localhost, and the SERVER_NAME if it's there in the .env file.
-ALLOWED_HOSTS = ['localhost', os.environ.get('SERVER_NAME', ''), '0.0.0.0', 'api.open5e.com', 'api-beta.open5e.com']
+ALLOWED_HOSTS = ['localhost', os.environ.get('SERVER_NAME', ''), '0.0.0.0', 'api.open5e.com', 'api-beta.open5e.com', '127.0.0.1']
```
Comment (Dashron) [inline server/settings.py:33]: This was one of the changes I needed to get the server running locally

File: api/views.py [hunk 3]
```diff
@@ -17,14 +18,27 @@ class ManifestViewSet(viewsets.ReadOnlyModelViewSet):
...
 
 class SearchView(HaystackViewSet):
     """
     API endpoint that allows searching our database.
     """
-
+    schema = CustomSchema(
+        summary={
+			'/search/': 'Search',
+			'/search/{id}/': 'Search', # I doubt this is a real endpoint
```
[the remark this answers is not in the document]
Comment (eepMoody) [reply api/views.py:38]: Ah, you are correct! Yes, feel free to remove it.

File: api/serializers.py (hunk 1 above)
Comment (eepMoody) [reply api/serializers.py:39]: Huh, I can't imagine why it removes them, maybe they're unnecessary?

Let's try it out in Staging and see what happens. :)

**Review by eepMoody — APPROVED** [review]

File: api/views.py [hunk 4]
```diff
@@ -177,6 +239,13 @@ class ConditionViewSet(viewsets.ReadOnlyModelViewSet):
     """
     API endpoint that allows viewing of Conditions.
     """
+    schema = CustomSchema(
+        summary={
+			'/conditions/': 'View Conditions',
+			'/conditions/{slug}/': 'View Condition',
```
Comment (donjonlord) [inline api/views.py:245]:
```suggestion
			'/conditions/{slug}/': 'Retrieve Condition',
```

File: api/views.py [hunk 5]
```diff
@@ -188,6 +257,13 @@ class RaceViewSet(viewsets.ReadOnlyModelViewSet):
     """
     API endpoint that allows viewing of Races and Subraces.
     """
+    schema = CustomSchema(
+        summary={
+			'/races/': 'View Races',
```
Comment (donjonlord) [inline api/views.py:262]:
```suggestion
			'/races/': 'List Races',
```

File: api/views.py [hunk 6]
```diff
@@ -188,6 +257,13 @@ class RaceViewSet(viewsets.ReadOnlyModelViewSet):
...
+			'/races/{slug}/': 'View Race',
```
Comment (donjonlord) [inline api/views.py:263]:
```suggestion
			'/races/{slug}/': 'Retrieve Race',
```

File: api/views.py [hunk 7]
```diff
@@ -199,6 +275,13 @@ class SubraceViewSet(viewsets.ReadOnlyModelViewSet):
     """
     API endpoint that allows viewing of Races and Subraces.
     """
+    schema = CustomSchema(
+        summary={
+			'/subraces/': 'View Subraces',
```
Comment (donjonlord) [inline api/views.py:280]:
```suggestion
			'/subraces/': 'List Subraces',
```

File: api/views.py [hunk 8]
```diff
@@ -199,6 +275,13 @@ class SubraceViewSet(viewsets.ReadOnlyModelViewSet):
...
+			'/subraces/{slug}/': 'View Subrace',
```
Comment (donjonlord) [inline api/views.py:281]:
```suggestion
			'/subraces/{slug}/': 'Retrieve Subrace',
```

File: api/views.py [hunk 9]
```diff
@@ -210,6 +293,13 @@ class CharClassViewSet(viewsets.ReadOnlyModelViewSet):
     """
     API endpoint that allows viewing of Classes and Archetypes.
     """
+    schema = CustomSchema(
+        summary={
+			'/classes/': 'View Classes',
```
Comment (donjonlord) [inline api/views.py:298]:
```suggestion
			'/classes/': 'List Classes',
```

File: api/views.py [hunk 10]
```diff
@@ -232,6 +329,13 @@ class MagicItemViewSet(viewsets.ReadOnlyModelViewSet):
     """
     API endpoint that allows viewing of Magic Items.
     """
+    schema = CustomSchema(
+        summary={
+			'/magicitems/': 'View Magic Items',
```
Comment (donjonlord) [inline api/views.py:334]:
```suggestion
			'/magicitems/': 'List Magic Items',
```

File: api/views.py [hunk 11]
```diff
@@ -232,6 +329,13 @@ class MagicItemViewSet(viewsets.ReadOnlyModelViewSet):
...
+			'/magicitems/{slug}/': 'View Magic Item',
```
Comment (donjonlord) [inline api/views.py:335]:
```suggestion
			'/magicitems/{slug}/': 'Retrieve Magic Item',
```

File: api/views.py [hunk 12]
```diff
@@ -244,6 +348,13 @@ class WeaponViewSet(viewsets.ReadOnlyModelViewSet):
     """
     API endpoint that allows viewing of Weapons.
     """
+    schema = CustomSchema(
+        summary={
+			'/weapons/': 'View Weapons',
```
Comment (donjonlord) [inline api/views.py:353]:
```suggestion
			'/weapons/': 'List Weapons',
```

File: api/views.py [hunk 13]
```diff
@@ -244,6 +348,13 @@ class WeaponViewSet(viewsets.ReadOnlyModelViewSet):
...
+			'/weapons/{slug}/': 'View Weapon',
```
Comment (donjonlord) [inline api/views.py:354]:
```suggestion
			'/weapons/{slug}/': 'Retrieve Weapon',
```

File: api/views.py [hunk 14]
```diff
@@ -256,6 +367,13 @@ class ArmorViewSet(viewsets.ReadOnlyModelViewSet):
     """
     API endpoint that allows viewing of Armor.
     """
+    schema = CustomSchema(
+        summary={
+			'/armor/': 'View Armor',
```
Comment (donjonlord) [inline api/views.py:372]:
```suggestion
			'/armor/': 'List Armor',
```

File: api/views.py [hunk 15]
```diff
@@ -256,6 +367,13 @@ class ArmorViewSet(viewsets.ReadOnlyModelViewSet):
...
+			'/armor/{slug}/': 'View Armor',
```
Comment (donjonlord) [inline api/views.py:373]:
```suggestion
			'/armor/{slug}/': 'Retrieve Armor',
```

File: openapi-schema.yml [hunk 16]
```diff
@@ -0,0 +1,3694 @@
...
+                          format: date-time
+                          readOnly: true
+                      required:
+                      - filename
+                      - type
+                      - hash
+          description: ''
+      summary: View Manifests
+      tags: &id001
+      - Manifest
```
Comment (donjonlord) [inline openapi-schema.yml:89]: Plural?

**Review by donjonlord — COMMENTED** [review]
Nice. I might have gone too far in my interpretation of what constitutes text that can easily change vs. text that, if you change it, messes things up. If any of these suggestions are in the second category, please feel free to disregard.

**Review by donjonlord — COMMENTED** [review]
Nice. I might have gone too far in my interpretation of what constitutes text that can easily change vs. text that, if you change it, messes things up. If any of these suggestions are in the second category, please feel free to disregard.

File: api/serializers.py (hunk 1 above)
Comment (Dashron) [reply api/serializers.py:39]: oh it's not clear from the other diff, but I just was talking about the trailing spaces. there seems to be a handful around other files too.

File: openapi-schema.yml [hunk 17]
```diff
@@ -0,0 +1,3694 @@
+openapi: 3.0.2
+info:
+  title: Open 5e
```
[the remark this answers is not in the document]
Comment (Dashron) [reply openapi-schema.yml:3]: it won't, fixed!

**Comment by Sturlen** [comment]
Would it make sense to regenerate the OAS as part of the build step, instead of committing it? Otherwise you have to remember to run it manually each time the schema changes. You then run the risk of the schema being out of sync with the server.

**Comment by eepMoody** [comment]
@Sturlen it looks like the [OpenAPI official github action](https://docs.readme.com/main/docs/github-actions-openapi-example) expects you to be checking in the OAS definition file.

I feel like this is a solid start, then we can figure out how to automate OAS auto-builds as part of development or that GH action once it's up and we can see the output?

**Review by dprothero — COMMENTED** [review]
Excited for this!

**Comment by Dashron** [comment]
We definitely would want to run the script as part of the build process, and I like committing the file because then it's easy to find and check and download in your browser

**Pull request merged**